src/devtools/client/debugger/src/components/Editor/Breakpoints/useBreakpoints.ts

```diff
@@ -0,0 +1,113 @@
+import { PointsContext } from "bvaughn-architecture-demo/src/contexts/PointsContext";
+import {
+  getDocument,
+  resizeBreakpointGutter,
+  toEditorLine,
+} from "devtools/client/debugger/src/utils/editor";
+import type { SourceEditor } from "devtools/client/debugger/src/utils/editor/source-editor";
+import { features } from "devtools/client/debugger/src/utils/prefs";
+import { useContext, useEffect } from "react";
+import { getSelectedSource } from "ui/reducers/sources";
+import { useAppSelector } from "ui/setup/hooks";
+import classNames from "classnames";
+
+const { codeFolding } = features;
+
+export default function useEditorBreakpoints(sourceEditor: SourceEditor | null) {
+  const selectedSource = useAppSelector(getSelectedSource);
+
+  const { deletePoints, editPoint, points } = useContext(PointsContext);
+
+  useEffect(() => {
+    if (!points || !selectedSource || !sourceEditor) {
+      return;
+    }
+
+    const { editor } = sourceEditor;
+
+    const doc = getDocument(selectedSource.id);
+    if (!doc) {
+      return;
+    }
+
+    // We only need to render one gutter breakpoint per line.
+    // This Set is used to track which lines we've already rendered.
+    const renderedLines = new Set();
+
+    points.forEach(point => {
+      if (!point.shouldBreak) {
+        return;
+      }
+
+      const lineNumber = point.location.line;
+      if (renderedLines.has(lineNumber)) {
+        return;
+      }
+
+      renderedLines.add(lineNumber);
+
+      const marker = makeMarker();
```

Comment:
Yes.

This was kind of a conscious decision, since it provided a simple way for adding vs removing. (I wanted to avoid having to manually diff the previous and next `points` array to determine what needed to be added or removed.)

I _think_ it's probably fine in practice since there aren't many points, as you say.